proxy/src/fs_watch.rs

```diff
@@ -0,0 +1,703 @@
+            test: fn(Self, Watch<()>, Box<Future<Item=(), Error = ()>>)
+        ) {
+            let stream = stream_changes_polling(
+                self.paths.clone(),
+                Duration::from_secs(1)
+            );
+            let (watch, bg) = watch_stream(stream);
+            test(self, watch, bg)
+        }
+
+        #[cfg(target_os="linux")]
+        fn test_inotify(
+            self,
+            test: fn(Self, Watch<()>, Box<Future<Item=(), Error = ()>>)
+        ) {
+            let paths = self.paths.clone();
+            let stream = inotify::WatchStream::new(paths)
+                .unwrap()
+                .map_err(|e| panic!("{}", e));
+            let (watch, bg) = watch_stream(stream);
+            test(self, watch, bg)
+        }
+    }
+
+    fn create_file<P: AsRef<Path>>(path: P) -> io::Result<File> {
+        let f = File::create(path)?;
+        println!("created {:?}", f);
+        Ok(f)
+    }
+
+    fn create_and_write<P: AsRef<Path>>(path: P, s: &[u8]) -> io::Result<File> {
+        let mut f = File::create(path)?;
+        f.write_all(s)?;
+        println!("created and wrote to {:?}", f);
+        Ok(f)
+    }
+
+    fn watch_stream(stream: impl Stream<Item = (), Error = ()> + 'static)
+        -> (Watch<()>, Box<Future<Item = (), Error = ()>>)
+    {
+        let (watch, store) = Watch::new(());
+        // Use a watch so we can start running the stream immediately but also
+        // wait on stream updates.
+        let f = stream
+            .forward(store.sink_map_err(|_| ()))
+            .map(|_| ())
+            .map_err(|_| ());
+
+        (watch, Box::new(f))
+    }
+
+    fn next_change(rt: &mut Runtime, watch: Watch<()>)
+        -> Result<(Option<()>, Watch<()>), WatchError>
+    {
+        let next = watch.into_future().map_err(|(e, _)| e);
+        rt.block_on_for(Duration::from_secs(2), next)
+    }
+
+    fn test_detects_create(
+        fixture: Fixture,
+        watch: Watch<()>,
+        bg: Box<Future<Item = (), Error = ()>>
+    ) {
+        let Fixture { paths, dir: _dir, mut rt } = fixture;
+
+        rt.spawn(bg);
+
+        paths.iter().fold(watch, |watch, path| {
+            create_file(path).unwrap();
+
+            let (item, watch) = next_change(&mut rt, watch).unwrap();
+            assert!(item.is_some());
+            watch
+        });
+    }
+
+    fn test_detects_delete_and_recreate(
+        fixture: Fixture,
+        watch: Watch<()>,
+        bg: Box<Future<Item = (), Error = ()>>
+    ) {
+        let Fixture { paths, dir: _dir, mut rt } = fixture;
+        rt.spawn(bg);
+
+        let watch = paths.iter().fold(watch, |watch, ref path| {
+            create_and_write(path, b"A").unwrap();
+
+            let (item, watch) = next_change(&mut rt, watch).unwrap();
+            assert!(item.is_some());
+            watch
+        });
+
+        for path in &paths {
+            fs::remove_file(path).unwrap();
+            println!("deleted {:?}", path);
```

Comment:
It seems like here we need `assert!(item.is_some());` too. And maybe other places?

Reply:
Sorry, what do you mean? I don't see an `item` in scope there?

Reply:
I mean, shouldn't there be something like this here:

```rust
let (item, watch) = next_change(&mut rt, watch).unwrap();
assert!(item.is_some());
```

Otherwise, how do we know if the notifications received in the `next_change()` calls below are from the `create_and_write` and not from the `remove_file`?

Reply:
Ah, yes, you're right; when I rewrote this test, I think I must've briefly forgot that the filtering out of deletions happens in the TLS-specific code and not here. Will fix.